README.md
# F-Seq2: a feature density estimator for high-throughput sequence tags

Tag sequencing using high-throughput sequencing technologies are employed to identify specific sequence features such as 
DNase-seq, ATAC-seq, ChIP-seq, and FAIRE-seq. To intuitively summarize and display individual sequence data as an 
accurate and interpretable signal, we have developed the original [F-Seq](http://fureylab.web.unc.edu/software/fseq/) 
[GitHub](https://github.com/aboyle/F-seq), a software package that generates a continuous tag sequence density 
estimation allowing identification of biologically meaningful sites whose output can be displayed directly in the UCSC 
Genome Browser. 

F-Seq2 is a complete rewrite and improvement of the original version in Python. F-Seq2 implements a dynamic parameter to 
conduct statistical analysis in “continuous” Poisson distribution. Combining the dynamic test with the 
statistically rigorous read distribution modeled by KDE allows us to more robustly account for local bias and solve ties 
when ranking candidate summits. F-Seq2 reconstructs a more biologically accurate signal by weighting the reads with 
user-input control data. A higher quality signal leads to better sanity-check data, compare and combine signals 
from different datasets.



## Table of contents

1. [Installation](./INSTALL.md)
2. [Usage](#usage)
    - [`callpeak`](#callpeak)
    - [`callpeak_idr`](#callpeak_idr)
    - [`idr`](#idr)
3. [Output files and formats](#output-files-and-formats)
4. [Examples](#examples)
5. [Reference](#reference)
6. [Troubleshooting](#troubleshooting)



## Installation
See [here](./INSTALL.md).



## Usage

```
fseq2 [-h] [--version]
    {callpeak, callpeak_idr, idr}
```

Available subcommands

Subcommand | Description
-----------|----------
`callpeak` | F-Seq2 main function to call peaks from alignment results.
`callpeak_idr` | Call peaks and follow by IDR framework with recommended settings.
`idr` | A wrapper for [IDR package](https://github.com/nboley/idr) for customized IDR analysis.



## `callpeak`
#### Command line input:
##### `-treatment_file`
REQUIRED argument for fseq2. Treatment file(s) in bam or bed format. If specifiy multiple files (separated by space), 
they are considered as one treatment experiment. See [here](./INPUT_FORMAT.md) for more details about input format.

##### `-control_file`
Control file(s) corresponding to treatment file(s).

##### `-pe`
Paired-end mode. If this flag on, treatment (and control) file(s) are paired-end data, either in format of BAMPE or BEDPE. 
Default is False to treat all data as single-end. See [here](./INPUT_FORMAT.md) for more details about paired-end mode.

##### `-chrom_size_file` 
A file specify chrom sizes, where each line has one chrom and its size. This is required if output signal format is bigwig. 
Note if this file is specified, fseq2 only process the chroms in this file. Default is False to process all and cannot output bigwig.

##### `-o`
Output directory. Default is current directory.

##### `-name`
Prefix for all output files. This overrides exisiting files. Default is `fseq2_result`.

##### `-sig_format`
Signal format for reconstructed signal. Available format `wig`, `bigwig`, `np_array`. Note if choose `np_array`, arrays 
for each chrom are stored in [`NAME_sig.h5`](#name_sigh5) with `chrom` as key, and no gaussian smooth applied. Default is False, without output signal.

##### `-sort_by`
Sort peaks and summits by `pValue` or `chromAndStart`. Default is `chromAndStart`.

##### `-v`
Verbose output. Default is False.  

##### `-f`
Fragment size of treatment data. Default is to estimate from data. This determines shift size where `offset = fragment_size/2`. 
For DNase-seq and ATAC-seq data, set `-f 0`. 

##### `-l`
Feature length for treatment data. Default is 600. Recommend 50 for TF ChIP-seq, 600 for DNase-seq and ATAC-seq, 
1000 for histone ChIP-seq.

##### `-fc`
Fragment size of control data.

##### `-t`
Threshold (standard deviations) to call candidate summits. Default is 4.0. Recommend 4.0 for broad peaks, 
8.0 for sharp peaks.

##### `-p_thr`
P value threshold. Default is 0.01.

##### `-q_thr`
Q value (FDR) threshold. Default is not set and use `p_thr`. If set, only use `q_thr`.

##### `-cpus`
Number of cpus to use. Default is 1.

##### `-nfr_upper_limit`
Nucleosome free region upper limit. Default is 150. Used as window_size and min_distance when `-f 0`.

##### `-pe_fragment_size_range`
Effective only if `-pe` on. Keep PE fragments whose size within the range. Default is False, without any selection. Useful for ATAC-seq data:  
(1) to call peaks on nucleosome free regions, specify: 0 150  
(2) to call peaks on nucleosome centers, specify: 150 inf

##### `-prior_pad_summit`
Prior knowledge about peak length which only padded into `NAME_summits.narrowPeak`. Default is 0. 
Useful for IDR analysis: in `callpeak_idr`, we set it to est. fragment size.

##### `-num_peaks`
Maximum number of peaks called. Default is not set. If set, overrides `p_thr` and `q_thr`.



## `callpeak_idr`
#### Command line input:
Most arguments are shared between `callpeak` and `callpeak_idr`. Here are the unique ones.  
> Notice if it is `-` or `--` ahead of arguments. `--` arguments are from IDR package. `-` are from fseq2.
##### `-treatment_file_1`
Treatment file in bam or bed format as replicate 1.

##### `-treatment_file_2`
Treatment file in bam or bed format as replicate 2.

##### `-control_file_1`
Control file in bam or bed format, paired with replicate 1 treament file.

##### `-control_file_2`
Control file in bam or bed format, paired with replicate 2 treament file.

##### `-name_1`
Prefix for output files for replicate 1 (default=`fseq2_result_1`).

##### `-name_2`
Prefix for output files for replicate 2 (default=`fseq2_result_2`).

##### `-prior_pad_summit`
Prior knowledge about peak length which only padded into `NAME_summits.narrowPeak`. Default is est. fragment size.

##### `--idr_threshold`
Only return peaks with a global idr threshold below this value. Default: report all peaks.

##### `--soft_idr_threshold`
Report statistics for peaks with a global idr below this value but return all peaks with an idr below --idr Default: 0.05.

##### `--plot`
Plot IDR results. Specify False if no plot. Default is to plot to `NAME_1_NAME_2.png`. Can specify other name here. 
Notice this is different from original IDR package which is only a flag.



## `idr`
#### Command line input and output:
See original [IDR documentation](https://github.com/nboley/idr#usage).  
> Notice all single letter arguments are removed to avoid conflict with fseq2, e.g. no `-s`, use `--samples`



## Output files and formats
#### `NAME_summits.narrowPeak` 
BED6+4 format
1. chrom
2. chromStart 
3. chromEnd 
4. name - `NAME_summit_num`, num is sorted by either `Pvalue` or `chromAndStart`.
5. score - `int(10*-log10(pValue))`.
6. strand - `.`
7. signalValue - Average treatment signal value given window size.
8. pValue - Measurement of statistical significance (-log10). Use -1 if no pValue is assigned.
9. qValue - Measurement of statistical significance using false discovery rate (-log10). Use -1 if no qValue is assigned.
10. peak - 0 if no specification of `-prior_pad_summit`.
  
#### `NAME_peaks.narrowPeak` 
Similar to summit file except that it can contain multiple summits information. 
For 7-10 columns, if multiple summits in a peak, output a comma separated list for each column.
1. chrom
2. chromStart 
3. chromEnd 
4. name - `NAME_peak_num`, num is sorted by either `Pvalue` or `chromAndStart`.
5. score - Max `int(10*-log10(pValue))` of all summits.
6. strand - `.`
7. signalValue
8. pValue
9. qValue
10. peak - Relative summit position(s) to peak start.

#### `NAME.bw` and `NAME.wig`
Reconstructed signal files which can be displayed directly in the UCSC Genome Browser. 
Recommend `bw` for efficient indexing in the browser.  

#### `NAME_sig.h5`
Reconstructed signal file without any smoothing. Signal is stored for each chrom in `np.array` and accessed by key `chrom`.   
For example:
```
>>> with h5py.File(NAME_sig.h5, mode='r') as sig_file:
...     signal = sig_file['chr1'][:] # read in all signal on chr1
```

#### `NAME_1_NAME_2_conservative_IDR_thresholded_peaks.narrowPeak` and `NAME_1_NAME_2.png`
Generated by `fseq2 callpeak_idr`.    Detailed format information is [here](https://github.com/nboley/idr#output).



## Examples

#### DNase-seq data
```
$ fseq2 callpeak treatment_file.bam -f 0 -l 600 -t 4.0 -v -cpus 5
```

#### ATAC-seq data
Paired-end ATAC-seq data, and call peaks on open chromatin regions, without calling on nucleosomes
```
$ fseq2 callpeak treatment_file.bam -f 0 -l 600 -t 4.0 -pe -nfr_upper_limit 150 -pe_fragment_size_range 0 150
```

#### ChIP-seq data
TF ChIP-seq data
```
$ fseq2 callpeak treatment_file.bed -control_file control_file.bed -l 50 -t 8.0 -sig_format bigwig -chrom_size_file ./tests/integration/fixtures/hg19.chrom.sizes  
```



## Troubleshooting

##### 1. Install error on mac Mojave: 
```
fatal error: 'ios' file not found 
#include "ios"
```
Solution:  
add `CFLAGS='-stdlib=libc++'` in front of `pip install`
```
$ CFLAGS='-stdlib=libc++' pip install fseq2
```

##### 2. Memory error

Solution:  
try with less CPUs


##### 3. `NotImplementedError: "bamToBed" does not appear to be installed`

Solution:  
update bedtools >= 2.29.0


##### 4. Warnings when `-pe`
Mostly likely bam file is not sorted by name.  
Solution:  
see [here](./INPUT_FORMAT.md)

##### 5. Too few peaks after multi-test correction
This may indicate poor data quality.  
Solution:  
use `-p_thr` instead of `-q_thr`


HISTORY.rst
=======
History
=======

2.0 (2020-08-03)
------------------

* First release on PyPI.


fseq.py
#!urs/bin/env python3

"""F-Seq Version 2.0 Prototype

"""

__author__ = 'Nanxiang(Samuel) Zhao'
__email__ = 'samzhao@umich.edu'
__version__ = 2.0

import argparse
import sys
import functools
import pandas as pd
import pybedtools
import numpy as np
import KDEpy
import pyBigWig
from itertools import chain
import multiprocessing as mp
import time
import math
import os


def calculate_fragment_size(cuts_df, verbose=False):
    """Estimate fragment size using the chr with most cuts.
    This estimation is done by conducting Wilcoxon-Mann-Whitney rank sum test. These fragments are usually < 200bp
    so we can assume that the average is < 500.

    Input:
        cuts_df: (pd.Dataframe) the cuts dataframe with most cuts, must have cols =
        ['chrom', 'start', 'end', 'strand', 'cuts']

    Return:
        fragment_size: (int) estimated fragment size
    """
    max_range = 500 # max search window for each + cut
    prior = 150
    max_cut_to_use = 1000000 #1 million is a good estimate
    if max_cut_to_use > cuts_df.shape[0] - max_range:
        max_cut_to_use = cuts_df.shape[0] - max_range

    cuts_df_mod = cuts_df[:max_cut_to_use].copy()
    pos_cuts_ls = cuts_df_mod[cuts_df_mod['strand'] == '+']['cuts'].values
    cuts_ls = cuts_df_mod['cuts'].values
    neg_strand_ls = (cuts_df_mod['strand'] == '-').values

    if verbose:
        print('Calculating fragment size (this may take ~ 3 min, specify -f 0 if using DNase HS data) ...',
              flush=True)

    fragment_size = int(np.median(np.concatenate([cuts_ls[(cuts_ls > (current_cuts - max_range + prior)) &
                                                          (cuts_ls < (current_cuts + max_range + prior)) &
                                                          neg_strand_ls] - current_cuts
                                                  for current_cuts in pos_cuts_ls], axis=0)))
    # todo:
    #   1. can speed up by multiprocessing
    #   2. consider use filter() to speed up

    if verbose:
        print('Done!', flush=True)

    return fragment_size


def calculate_bandwidth(feature_length):
    """Calculate bandwidth.
    """
    bandwidth = float(feature_length / 2 / 3) # 3 standard deviations

    return bandwidth


def calculate_threshold(size, ncuts, std, bandwidth):
    """Calculate threshold for peak calling.

    Input:
        size: sum range of cuts across all chroms in bp
        ncuts: total number of cuts
        std: standard deviation for calculating threshold
        window_size: window size in bp, window_size = int(2860/2) when bandwidth = 100

    Return:
        threshold: threshold for peak calling
    """
    # calculate window_size
    window_size = 1400.0; java_min_value = 2 ** (-149)
    while True:
        window_size += 1
        x = window_size / bandwidth
        v = math.exp(-(x * x) / 2) / math.sqrt(math.pi * 2)
        if v < java_min_value:
            break
    window_size = int(window_size - 1)

    total_window = 1 + window_size * 2
    cut_density = int((ncuts / size) * total_window)
    threshold_iterations = 10000

    np.random.seed(934) # may not need it
    simulated_results = [KDEpy.FFTKDE(bw=bandwidth).fit(np.random.randint(total_window, size=cut_density)).evaluate(
        np.arange(-1, total_window + 1))[window_size + 1] for _ in range(threshold_iterations)]
    # the following has more readability
    #simulated_results = []
    #for _ in range(threshold_iterations):
        #simulated_cuts = np.random.randint(total_window, size=cut_density)
        #kdepy_kde = KDEpy.FFTKDE(bw=bandwidth)
        #kdepy_kde.fit(simulated_cuts)
        #simulated_results.append(kdepy_kde.evaluate(np.arange(-1, total_window + 1))[window_size + 1])
    kdepy_result = (np.array(simulated_results) * 20000000) / ncuts * cut_density

    threshold = np.mean(kdepy_result) + std * np.std(kdepy_result)

    return threshold


def run_kde(cuts_df, chrom, bf, pf, fragment_offset, bandwidth, ncuts, verbose=False):
    """Estimate KDE.

    Return:
        [chrom, first_cut, kde_results]: a list contains chrom name, first cut position, and kde results from first cut
            to last cut in np.array
    """
    start_time = time.time()
    if bf:
        with pyBigWig.open(bf) as background_bw:
            background_bw_weights = np.array(list(chain.from_iterable(
                [background_bw.values(chrom, start, start + 1) * duplicate
                 for start, duplicate in zip(np.unique(cuts_df['start'].values),
                                             cuts_df.groupby('start')['start'].count().values)])))
            # bigwig is 0-index, wig is 1-index
        # 1/background_bw_weights except 0 which remains 0
        background_bw_weights = np.divide(np.ones(background_bw_weights.shape[0]), background_bw_weights,
                                          out=np.zeros_like(background_bw_weights), where=(background_bw_weights != 0))
        #background_bw_weights[background_bw_weights == 0] = 1 / 5
        #background_bw_weights = 1 / background_bw_weights
    if pf:
        with pyBigWig.open(pf) as ploidy_bw:
            ploidy_bw_weights = np.array(list(chain.from_iterable(
                [ploidy_bw.values(chrom, start, start + 1) * duplicate
                 for start, duplicate in zip(np.unique(cuts_df['start'].values),
                                             cuts_df.groupby('start')['start'].count().values)])))
        ploidy_bw_weights = ploidy_bw_weights / 1000

    if bf and pf:
        weights = background_bw_weights / ploidy_bw_weights
    elif bf:
        weights = background_bw_weights
    elif pf:
        weights = 1 / ploidy_bw_weights
    else:
        weights = np.ones(cuts_df.shape[0])

    cuts = cuts_df['cuts'].values
    first_cut = cuts.min()
    last_cut = cuts.max()
    if fragment_offset == 0:
        kdepy_kde = KDEpy.FFTKDE(bw=bandwidth).fit(cuts, weights=weights)
        kdepy_result = kdepy_kde.evaluate(np.arange(first_cut - 1, last_cut + 2))[1:-2].astype(np.float16)
        #kdepy_result = (kdepy_result * 20000000) / ncuts * (weights.sum())
    else:
        cuts_w_offset = ((cuts_df['start'] + fragment_offset) * (cuts_df['strand'] == '+') +
                         (cuts_df['end'] - fragment_offset) * (cuts_df['strand'] == '-')).astype('int').values
        kde_evaluation_start = min(cuts_w_offset.min(), first_cut); kde_evaluation_end = max(cuts_w_offset.max(), last_cut)
        #assert (first_cut - kde_evaluation_start) >= 0
        kdepy_kde = KDEpy.FFTKDE(bw=bandwidth).fit(cuts_w_offset, weights=weights)
        try:
            kdepy_result = kdepy_kde.evaluate(np.arange(kde_evaluation_start - 1, kde_evaluation_end + 2))[1:-2].astype(np.float16)
        except:
            np.fft.restore_all() # if encounter the bug for MLK https://github.com/IntelPython/mkl_fft/issues/24
            kdepy_result = kdepy_kde.evaluate(np.arange(kde_evaluation_start - 1, kde_evaluation_end + 2))[1:-2].astype(np.float16)
        kdepy_result = kdepy_result[(first_cut - kde_evaluation_start):(last_cut - kde_evaluation_start)]
        #kdepy_result = (kdepy_result * 20000000) / ncuts * (weights.sum()) / 2
        # this scaling is simulating results with offset (left window positive cuts and right window negative cuts) by
        # multiplying all signal with 0.5
    kdepy_result = (kdepy_result * 20000000) / ncuts * (weights.sum()) # lets use all reads, more related to threshold
    end_time = time.time()

    if verbose:
        print(f'{chrom}: first={first_cut}, last={last_cut} \n'
              f'{chrom}: Completed in {end_time - start_time:.3f} seconds. \n'
              f'------------------------------------- \n', flush=True)

    # todo:
    #   1. F-seq 1.85 timing includes output wig. Output bw takes longer time and we do not measure it here.
    #   We also need to measure the time for calling peaks.

    return [chrom, first_cut, kdepy_result]

def chrom_sizes_reader(file_name):
    """ Chromosome size reader, used for genearating bigwig file.

    """
    chr_size_dic = {}
    with open(file_name) as f:
        for line in f:
            (key, val) = line.split()
            chr_size_dic[key] = int(val)
    return chr_size_dic


if __name__ == '__main__':
    # command line input settings
    parser = argparse.ArgumentParser(
        description='This command executes fseq which is a feature density estimator for high-throughput sequence tags')
    parser.add_argument('input_file', help='input file(s) in bam or bed format. If multiple input files, separate '
                                           'by space; if one input file, it can contain all chromosomes', nargs='+')
    parser.add_argument('-b', help='background directory (default=none)')
    parser.add_argument('-bf', help='background bigwig file for all chromosomes (default=none)')
    parser.add_argument('-f', help="fragment size (default=estimated from data). For DNase HS data (5' ends)"
                                   " set -f 0", default=False, type=int)
    parser.add_argument('-l', help='feature length (default=600)', default=600, type=int)
    parser.add_argument('-o', help='output directory (default=current directory)')
    parser.add_argument('-of', help='output format bigwig, wig, bed, or npf (default bigwig)')
    parser.add_argument('-p', help='ploidy/input directory (default=none)')
    parser.add_argument('-pf', help='ploidy/input bigwig file for all chromosome (default=none)')
    parser.add_argument('-s', help='wiggle track step (default=1)', default=1, type=int)
    parser.add_argument('-t', help='threshold (standard deviations) (default=4.0)', default=4.0, type=float)
    parser.add_argument('-v', help='verbose output', default=False, action='store_true')
    parser.add_argument('-cpus', help='number of cpus to use (default=min(unique_chrs, all-2))', default=False, type=int)
    # related to output
    parser.add_argument('-cs', help='chromosome size file, used for generating bigwig file (default=hg19)', default='hg19')
    parser.add_argument('-name', help='prefix for output files (default=fseq_run)', default='fseq_run')

    args = parser.parse_args()


    ### 1. Read in input files ###
    if args.input_file[0][-3:] == 'bam':
        input_bed = pd.concat([pybedtools.BedTool(file).bam_to_bed().to_dataframe() for file in args.input_file],
                              axis=0)
    elif args.input_file[0][-3:] == 'bed':
        input_bed = pd.concat(map(functools.partial(pd.read_csv, sep='\t', header=None), args.input_file),
                              axis=0).rename(columns={0:'chrom', 1:'start', 2:'end', 3: 'name', 4: 'score', 5:'strand'})
    else:
        sys.exit('Error: Input file format should be either .bam or .bed, please convert files and try again.')
    input_bed = input_bed.drop(['name', 'score'], axis=1)
    input_bed['cuts'] = input_bed['start'] * (input_bed['strand'] == '+') + \
                        input_bed['end'] * (input_bed['strand'] == '-')

    # TODO:
    #  1. check file type?
    #  2. check strand implicitly later
    #  3. maybe use dask to speed up calculations on dataframe, and solve if dataframe not fit in memory
    #  4. chrom size input for other genome build; this is only needed for bigwig output
    #  5. maybe only save useful info in np array to avoid memory issue, and not using dask


    ### 2. Set up all parameters ###
    if args.f is False:
        fragment_size = calculate_fragment_size(
            input_bed[input_bed['chrom'] == input_bed.groupby('chrom')['chrom'].count().idxmax()], verbose=args.v)
    else:
        fragment_size = args.f
    fragment_offset = int(fragment_size / 2)
    feature_length = args.l
    bandwidth = calculate_bandwidth(feature_length)
    sequence_length = input_bed.iloc[0, 2] - input_bed.iloc[0, 1]
    ncuts = input_bed.shape[0]
    threshold = calculate_threshold(size=input_bed.groupby('chrom')['cuts'].agg(np.ptp).sum(), ncuts=ncuts,
                                    std=args.t, bandwidth=bandwidth)

    if args.v:
        print('=====================================', flush=True)
        print(f'F-Seq Version {__version__} Prototype', flush=True)
        print('=====================================', flush=True)
        print('Settings: ', flush=True)
        print(f'\tbandwidth = {bandwidth}', flush=True)
        print(f'\tthreshold = {threshold:.3f}', flush=True)
        print(f'\test. fragment size = {fragment_size}', flush=True)
        print(f'\tsequence length = {sequence_length}', flush=True)
        print('-------------------------------------', flush=True)

    # TODO:
    #   1. check if sequence(read)_length consistent
    #   2. both signal and background normalize using total number of cuts, be aware


    ### 3. Parallel kde
    chrom_ls = input_bed['chrom'].unique()
    input_param_ls = [[input_bed[input_bed['chrom'] == chrom]] + [chrom, args.bf, args.pf, fragment_offset, bandwidth,
                                                                  ncuts, args.v] for chrom in chrom_ls]
    if not args.cpus:
        cpus = min(chrom_ls.shape[0], mp.cpu_count() - 2)
    else:
        cpus = args.cpus

    with mp.Pool(processes=cpus) as pool:
        results = pool.starmap(run_kde, input_param_ls)


    ### 4. Output results
    #for result in results:
        #with pyBigWig.open(args.o+'/'+result[0]+'.bw', 'w') as output_bw:
            #output_bw.addHeader([(result[0], CHR_SIZE_DIC[result[0]])])
            #output_bw.addEntries(result[0], result[1], values=result[2], span=1, step=1)

    if args.cs == 'hg19':
        chr_size_dic = {'chr1': 249250621, 'chr2': 243199373, 'chr3': 198022430, 'chr4': 191154276, 'chr5': 180915260,
                        'chr6': 171115067, 'chr7': 159138663, 'chr8': 146364022, 'chr9': 141213431, 'chr10': 135534747,
                        'chr11': 135006516, 'chr12': 133851895, 'chr13': 115169878, 'chr14': 107349540, 'chr15': 102531392,
                        'chr16': 90354753, 'chr17': 81195210, 'chr18': 78077248, 'chr19': 59128983, 'chr20': 63025520,
                        'chr21': 48129895, 'chr22': 51304566, 'chrX': 155270560, 'chrY': 59373566}  # hg19
    else:
        chr_size_dic = chrom_sizes_reader(args.cs)

    # deal with output file name to avoid overwrite
    if args.name == 'fseq_run':
        run_num = 1
        while os.path.exists(f'{args.o}/fseq_run_{run_num}.bw'):
            run_num += 1
        output_file = f'{args.o}/fseq_run_{run_num}.bw'
    else:
        if not os.path.exists(f'{args.o}/{args.name}.bw'):
            output_file = f'{args.o}/{args.name}.bw'
        else:
            run_num = 1
            while os.path.exists(f'{args.o}/{args.name}_{run_num}.bw'):
                run_num+=1
            output_file = f'{args.o}/{args.name}_{run_num}.bw'

    with pyBigWig.open(output_file, 'w') as output_bw:
        output_bw.addHeader([(result[0], chr_size_dic[result[0]]) for result in results])
        for result in results:
            output_bw.addEntries(result[0], result[1], values=result[2], span=1, step=1)


    if args.v:
        print('Thank you for using F-seq.')

    # TODO:
    #   1. specify step
    #   2. avoid overwrite files (done)
    #   3. output name and in one file instead of separated by chrom (done)
    #   4. more chrom sizes included, such as hg38
